book/src/cross-program-invocation.md

```diff
@@ -0,0 +1,101 @@
+accounts, each owned by a separate on-chain program:
+
+```rust,ignore
+let message = Message::new(vec![
+    foo_instruction::foo(&alice_pubkey),
+    bar_instruction::bar(&bob_pubkey),
+]);
+client.send_message(&[&alice_keypair, &bob_keypair], &message);
+```
+
+The current implementation does not, however, allow the `bar` program to
+conveniently invoke `foo` instructions on the client's behalf:
+
+```rust,ignore
+let message = Message::new(vec![
+    bar_instruction::foobar(&alice_pubkey, &bob_pubkey),
+]);
+client.send_message(&[&alice_keypair, &bob_keypair], &message);
+```
+
+Currently, there is no way to create instruction `foobar` that executes executes
+`foo_instruction::foo` from the `bar` program. The workaround is to extend the
+`bar` program with the implementation of the `foo` program, and create `foo`
+accounts with `BAR_PROGRAM_ID`, which the `bar` program is permitted to modify.
+With that workaround, `bar` can modify foo-like accounts created by the `bar`
+program, but not foo accounts created by the `foo` program.
+
+
+## Proposed Solution
+
+The goal of this design is to modify Solana's runtime such that an on-chain
+program can invoke an instruction from other program.
+
+Given two on-chain programs `foo` and `bar`, each implementing instructions
+`foo` and `bar` respectively, we would like to implement bar as close as
+possible to as the following Rust code of a cross-module function call:
+
+```rust,ignore
+use foo;
+
+fn bar(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    ...
+}
+
+fn foobar(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    foo::foo(&keyed_accounts[1..])?;
+
+    bar(keyed_accounts)?;
+}
+```
+
+The above code would require that the `foo` crate be dynamically
+linked, so that a custom linker could could intercept calls and
+validate accesses to `keyed_accounts`. That is, even though the client
+intends to modify both `foo` and `bar` accounts, only `foo` program is
+permitted to modify the `foo` account, and only the `bar` program is
+permitted to modify the `bar` account.
+
+Backing off from that ideal cross-program call, a slightly more
+verbose solution to is expose foo's existing `process_instruction()`
+entrypoint to the bar program:
+
+```rust,ignore
+use foo_instruction;
+
+fn bar(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    ...
+}
+
+fn foobar(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    let alice_pubkey = keyed_accounts[1].key;
+    let instruction = foo_instruction::foo(&alice_pubkey);
+    process_instruction(&instruction)?;
+
+    bar(keyed_accounts)?;
+}
+```
+
+where `process_instruction()` is built into Solana's runtime and responsible
+for routing the given instruction to the `foo` program via the instruction's
+`program_id` field. Before invoking `foo()`, the runtime must also ensure
+that `bar` didn't modify any accounts owned by `foo`. Likewise, after invoking
+`foo()`, the runtime must ensure that `foo` didn't modify any accounts owned
+by `bar`.
+
```

Comment:
I think all the instruction validation checks need to be run, and I think before foo is called to validate bar, and after foo is called to validate foo.

Which is a bit awkward.  Bar would need to have preserved scarcity before calling foo.

Reply:
Yes, all the checks in multiple places. This is the part that Move does at compile-time.  We can do it at runtime and in exchange for that performance hit (that doesn't affect concurrency), don't have to ask devs to learn a new programming language.

Reply:
@aeyakovenko, thinking more on this, I can't think of an attack on `foo()` by mucking with bar-owned lamports before calling it.

Reply:
@garious seems dicey.  Since bar could muck with it, call foo then fix up the values and the runtime won’t fail the transaction.

Reply:
I agree, but they can do that with Account.data anyway.

Reply:
But foo was written with the assumption that when it’s reading bar data that it’s in a valid state.  We have to check that the memory is valid anyways, I think validating the lamports will be the cheap part.

Reply:
`bar` didn't exist on-chain at the time `foo` was deployed. Besides all that, Alice had to sign the transaction with the `bar` instruction.  If it maliciously convinces `foo()` to give up tokens, Alice authorized that.

Reply:
Foo may read from accounts or any program, even the ones loaded after.  They just need to have a compatible ABI.

It might not be possible for Alice to validate that every path that bar takes results in the outcome that Alice wants.  Unless we also require the expected instructions to be signed as well.

Reply:
@aeyakovenko your last message is a push for scripts vs. transitive signing for arbitrary instructions to foo?

Reply:
@rob-solana I am not sure it helps this problem.   I can’t think of cross domain “invoke” mechanisms that don’t pass the kernel and validate all the arguments.  lamports are a “kernel” resource, so when foo sees something from bar it needs to be valid. 

Basically, it should be safe to run a contest for an account in any program with the most lamports.

Reply:
Can we agree that running all `verify_instruction()` checks before and after `foo()` is the most conservative action? Let's just start there and consider relaxing checks in some future proposal.

Reply:
@aeyakovenko, @rob-solana, done.

Reply:
What this proposal says is that Alice trusts whatever bar::foobar() might ask foo to do with her foo state.  Is that what we want?

Reply:
after my last, I long for a rename:

* foo -> something that means "library" or "subordinate module"
* foobar() -> something that means "main" or controlller

Reply:
@rob-solana that’s what I am thinking for scripts.  Programs can’t call ‘process_instruction’, and scripts can’t mutate state directly.

Reply:
@rob-solana, "whatever bar::foobar() might ask foo to do" is misleading, and suggests that bar is mutable or a black box and therefore dangerous. It'd be more precise to say "exactly what bar::foobar() says for foo to do". Even more precise, "exactly what bar::foobar() says for foo to do within the capabilities of the foo program". That is, foobar() cannot modify her foo state directly.

Reply:
Another way of saying it, this proposal arguably relaxes what a transaction signature means to a program. Before and after this proposal, the signature indicates the account holder authorized the transaction. However, before this proposal, the account holder was only able to authorize transactions containing *explicit* references to instructions. After this proposal, those references are permitted to be *implicit*.  Anatoly's "Script" proposal keeps those references explicit, but with a high cost to programability. A copy of every instruction that might be invoked needs to be hoisted up into the transaction.

Reply:
@garious Ok, thanks.  I get it.  Let me think more about whether or not to worry.

What worries me today: most of what I think about in programming for Solana is making sure my program can't be tricked, that I've covered all of my state-transition bases.  My concerns are outside the scope of this proposal, but I think this proposal brings them into sharp relief.